File: src/main/kotlin/net/merayen/elastic/system/Action.java
package net.merayen.elastic.system;

public abstract class Action {
	protected ElasticSystem system;

	protected abstract void run();

	public synchronized void start(ElasticSystem es) {
		system = es;
		run();
	}

	protected void waitFor(Func func) {
		try {
			while(!func.run()) {
				system.update();
				synchronized (this) {
					wait(1);
				}
			}
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
	}

	public interface Func {
		boolean run();
	}
}

File: src/main/kotlin/net/merayen/elastic/system/Action.kt
package net.merayen.elastic.system

import net.merayen.elastic.backend.context.JavaBackend
import net.merayen.elastic.system.intercom.ElasticMessage

/**
 * An action is a group of messages sent into the backend, like initializing a new project, open an existing one etc.
 */
abstract class Action {
	interface Handler {
		/**
		 * Called whenever the Action wants to send messages to the backend.
		 */
		fun onMessage(message: ElasticMessage)

		/**
		 * Called when ElasticSystem can process messages.
		 */
		fun onUpdateSystem()
	}

	var handler: Handler? = null

	abstract fun run()

	/**
	 * Called whenever there are messages from backend
	 */
	abstract fun onMessageFromBackend(message: ElasticMessage)

	protected fun waitFor(func: () -> Boolean) {
		while (!func()) {
			handler?.onUpdateSystem()
			synchronized(this) {
				Thread.sleep(1)
			}
		}
	}

	/**
	 * Send messages to backend.
	 */
	protected fun send(message: ElasticMessage) = handler?.onMessage(message)
}
